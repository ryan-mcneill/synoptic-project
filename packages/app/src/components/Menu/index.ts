import Accordion from "./Accordion";
import InfoTile from "./InfoTile";
import PlaylistTab from "./PlaylistTab";
import SearchTab from "./SearchTab";
import SongList from "./SongList";
import SongMenu from "./SongMenu";
import SongsTab from "./SongsTab";
import TabButton from "./TabButton";
import Tabs from "./Tabs";

export {
  Accordion,
  InfoTile,
  PlaylistTab,
  SearchTab,
  SongList,
  SongMenu,
  SongsTab,
  TabButton,
  Tabs
};
