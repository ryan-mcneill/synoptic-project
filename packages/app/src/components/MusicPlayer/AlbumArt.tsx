import React, { ReactElement } from "react";
import { connect, ConnectedProps } from "react-redux";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faMusic } from "@fortawesome/free-solid-svg-icons";
import { State } from "../../store/types";

const mapStateToProps = (state: State) => ({
  albumId: state.playlist.current?.albums[0]._id
});

const connector = connect(mapStateToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

const AlbumArt: React.FC<PropsFromRedux> = ({ albumId }): ReactElement => {
  return (
    <div
      className="album-art"
      style={{
        width: "80vw",
        height: "80vw",
        position: "absolute",
        top: "42vh",
        left: "50vw",
        transform: "translate(-50%, -50%)",
        backgroundColor: "#c0c0c0",
        boxShadow: "0px 0px 64px 4px rgba(0,0,0,0.2)"
      }}
    >
      {albumId ? (
        <img
          alt="Album art"
          src={`/api/album/art/${albumId}`}
          width="100%"
          height="100%"
        />
      ) : (
        <div
          style={{
            fontSize: "400px",
            position: "absolute",
            top: "50%",
            left: "48%",
            transform: "translate(-50%, -50%)"
          }}
        >
          <FontAwesomeIcon icon={faMusic} />
        </div>
      )}
    </div>
  );
};

export default connector(AlbumArt);
