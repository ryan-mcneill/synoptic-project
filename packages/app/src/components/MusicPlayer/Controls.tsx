import React, { ReactElement, useState } from "react";
import { connect, ConnectedProps } from "react-redux";
import classnames from "classnames";
import ReactAudioPlayer from "react-audio-player";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faPause,
  faPlay,
  faStepBackward,
  faStepForward
} from "@fortawesome/free-solid-svg-icons";
import { nextSong, prevSong, togglePlaying } from "../../store/actions";
import { State } from "../../store/types";

const mapDispatchToProps = (dispatch: any) => ({
  togglePlaying: (playing?: boolean) => dispatch(togglePlaying(playing)),
  nextSong: () => dispatch(nextSong()),
  prevSong: () => dispatch(prevSong())
});

const mapStateToProps = (state: State) => ({
  isDark: state.currentSong.background?.isDark,
  isPlaying: state.currentSong?.isPlaying,
  songId: state.playlist.current?._id,
  data: state.playlist.current
});

const connector = connect(mapStateToProps, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

const Controls: React.FC<PropsFromRedux> = ({
  data,
  isDark,
  isPlaying,
  songId,
  togglePlaying,
  nextSong,
  prevSong
}): ReactElement => {
  const [songLength, setSongLength] = useState(0);
  const [songTime, setSongTime] = useState(0);
  let audioPlayer: ReactAudioPlayer;

  const onSongLoad = () => {
    if (audioPlayer?.audioEl?.current) {
      if (isPlaying) audioPlayer.audioEl.current.play();
      setSongLength(audioPlayer.audioEl.current.duration);
    }
  };

  const formatTime = (time: number): string => {
    const tempDate = new Date(Math.round(time) * 1000);
    const minutes = tempDate.getMinutes();
    const seconds = tempDate.getSeconds();

    return (
      minutes.toString().padStart(2, "0") +
      ":" +
      seconds.toString().padStart(2, "0")
    );
  };

  const playStatusHandler = () => {
    if (data && audioPlayer?.audioEl?.current) {
      togglePlaying();
      audioPlayer.audioEl.current[isPlaying ? "pause" : "play"]();
    }
  };

  const nextHandler = () => {
    if (data) {
      nextSong();
      setSongTime(0);
    }
  };

  const prevHandler = () => {
    if (data) {
      prevSong();
      setSongTime(0);
    }
  };

  return (
    <div
      style={{
        height: "10vh",
        width: "90vw",
        position: "absolute",
        bottom: 0,
        padding: "0 5vw 5vh 5vw"
      }}
    >
      <ReactAudioPlayer
        ref={(element: ReactAudioPlayer) => (audioPlayer = element)}
        src={songId ? `/api/song/${songId}` : ""}
        listenInterval={1000}
        onListen={(e) => setSongTime(e)}
        onLoadedMetadata={() => onSongLoad()}
        onEnded={() => nextHandler()}
      />
      <div
        style={{
          display: "flex",
          alignItems: "center",
          justifyContent: "center"
        }}
      >
        <div style={{ fontSize: "48px" }} onClick={() => prevHandler()}>
          <FontAwesomeIcon icon={faStepBackward} />
        </div>
        <div
          style={{ fontSize: "96px", margin: "0 96px" }}
          onClick={playStatusHandler}
        >
          <FontAwesomeIcon icon={isPlaying ? faPause : faPlay} />
        </div>
        <div style={{ fontSize: "48px" }} onClick={() => nextHandler()}>
          <FontAwesomeIcon icon={faStepForward} />
        </div>
      </div>
      <div
        style={{
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
          marginTop: "48px",
          fontSize: "32px"
        }}
      >
        <div>{formatTime(songTime)}</div>
        <input
          className={classnames({ dark: !isDark, light: isDark })}
          type="range"
          style={{ width: "100%", margin: "0 24px" }}
          min="0"
          max={songLength}
          defaultValue={songTime}
        />
        <div>{formatTime(songLength)}</div>
      </div>
    </div>
  );
};

export default connector(Controls);
