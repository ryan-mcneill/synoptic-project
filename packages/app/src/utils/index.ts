import { artistsToString } from "./artistsToString";
import { filterData } from "./filterData";
import { filterSongs } from "./filterSongs";
import { playlistsToArray } from "./playlistsToArray";

export { artistsToString, filterData, filterSongs, playlistsToArray };
