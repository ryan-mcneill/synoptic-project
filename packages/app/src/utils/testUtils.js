import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import reducers from "../store/reducers";

function render(
  ui,
  {
    initialState,
    store = createStore(reducers, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from "@testing-library/react";
export { render };
