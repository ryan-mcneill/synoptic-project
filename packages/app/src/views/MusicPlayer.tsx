import React, { ReactElement } from "react";
import { connect, ConnectedProps } from "react-redux";
import { AlbumArt, Controls, NowPlaying } from "../components/MusicPlayer";
import { State } from "../store/types";

const mapStateToProps = (state: State) => ({
  albumsLoading: state.albums.loading,
  artistsLoading: state.artists.loading,
  isDark: state.currentSong.background?.isDark,
  backgroundColor: state.currentSong.background?.colour
});

const connector = connect(mapStateToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

const MusicPlayer: React.FC<PropsFromRedux> = ({
  albumsLoading,
  artistsLoading,
  isDark,
  backgroundColor
}): ReactElement =>
  !albumsLoading && !artistsLoading ? (
    <div
      style={{
        display: "block",
        margin: 0,
        position: "absolute",
        height: "100%",
        width: "100%",
        color: isDark ? "white" : "black",
        fontSize: "18px",
        backgroundColor
      }}
    >
      <NowPlaying />
      <AlbumArt />
      <Controls />
    </div>
  ) : (
    <></>
  );

export default connector(MusicPlayer);
