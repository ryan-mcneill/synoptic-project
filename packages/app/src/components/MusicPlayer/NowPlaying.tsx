import React, { ReactElement } from "react";
import { connect, ConnectedProps } from "react-redux";
import { artistsToString } from "../../utils";
import { State } from "../../store/types";

const mapStateToProps = (state: State) => ({
  artists: state.playlist.current?.artists,
  songName: state.playlist.current?.name,
  data: state.playlist.current
});

const connector = connect(mapStateToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

const NowPlaying: React.FC<PropsFromRedux> = ({
  artists = [{ name: "Unknown Artist" }],
  songName = "Unknown",
  data
}): ReactElement => {
  return (
    <div
      style={{
        height: "5vh",
        width: "90vw",
        textAlign: "center",
        padding: "5vh 5vw 2vh 5vw"
      }}
    >
      <div style={{ fontSize: "24px" }}>
        {data ? "NOW PLAYING" : "NOTHING PLAYING"}
      </div>
      <div
        style={{
          fontSize: "48px",
          whiteSpace: "nowrap",
          overflow: "hidden",
          textOverflow: "ellipsis"
        }}
      >
        {data
          ? `${songName} - ${artistsToString(artists)}`
          : "Please make a song selection"}
      </div>
    </div>
  );
};

export default connector(NowPlaying);
