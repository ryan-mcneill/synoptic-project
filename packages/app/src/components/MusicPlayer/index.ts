import AlbumArt from "./AlbumArt";
import Controls from "./Controls";
import NowPlaying from "./NowPlaying";

export { AlbumArt, Controls, NowPlaying };
