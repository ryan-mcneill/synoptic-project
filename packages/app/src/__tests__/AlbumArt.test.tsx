import React from "react";
import { AlbumArt } from "../components/MusicPlayer";
import { render } from "../utils/testUtils";

describe("The AlbumArt component", () => {
  it("renders correctly when no image path is passed", () => {
    const { container } = render(<AlbumArt />, {
      initialState: { playlist: {} }
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders correctly when given an image path", () => {
    const { container } = render(<AlbumArt />, {
      initialState: { playlist: { current: { albums: [{ _id: "123456" }] } } }
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
