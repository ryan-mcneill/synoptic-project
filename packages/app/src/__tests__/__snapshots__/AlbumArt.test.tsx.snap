// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The AlbumArt component renders correctly when given an image path 1`] = `
<div
  class="album-art"
  style="width: 80vw; height: 80vw; position: absolute; top: 42vh; left: 50vw; transform: translate(-50%, -50%); background-color: rgb(192, 192, 192); box-shadow: 0px 0px 64px 4px rgba(0,0,0,0.2);"
>
  <img
    alt="Album art"
    height="100%"
    src="/api/album/art/123456"
    width="100%"
  />
</div>
`;

exports[`The AlbumArt component renders correctly when no image path is passed 1`] = `
<div
  class="album-art"
  style="width: 80vw; height: 80vw; position: absolute; top: 42vh; left: 50vw; transform: translate(-50%, -50%); background-color: rgb(192, 192, 192); box-shadow: 0px 0px 64px 4px rgba(0,0,0,0.2);"
>
  <div
    style="font-size: 400px; position: absolute; top: 50%; left: 48%; transform: translate(-50%, -50%);"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-music fa-w-16 "
      data-icon="music"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M470.38 1.51L150.41 96A32 32 0 0 0 128 126.51v261.41A139 139 0 0 0 96 384c-53 0-96 28.66-96 64s43 64 96 64 96-28.66 96-64V214.32l256-75v184.61a138.4 138.4 0 0 0-32-3.93c-53 0-96 28.66-96 64s43 64 96 64 96-28.65 96-64V32a32 32 0 0 0-41.62-30.49z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;
